File: javascript/12-god-object/test/account.test.js
import assert from 'node:assert';
import test from 'node:test';
import { Account } from '../src/account.js';

test('logs in with correct password', () => {
  const acc = new Account('a@b.com', 'secret');
  assert.strictEqual(acc.login('secret'), true);
});

test('rejects wrong password', () => {
  const acc = new Account('a@b.com', 'secret');
  assert.strictEqual(acc.login('wrong'), false);
});

test('logs out', () => {
  const acc = new Account('a@b.com', 'secret');
  assert.strictEqual(acc.logout(), true);
});

test('updates profile', () => {
  const acc = new Account('a@b.com', 'secret');
  const result = acc.update_profile('Alice', 'Developer');
  assert.strictEqual(result.name, 'Alice');
  assert.strictEqual(result.bio, 'Developer');
});

test('changes password', () => {
  const acc = new Account('a@b.com', 'secret');
  assert.strictEqual(acc.change_password('secret', 'new'), true);
  assert.strictEqual(acc.login('new'), true);
});

test('rejects bad old password', () => {
  const acc = new Account('a@b.com', 'secret');
  assert.strictEqual(acc.change_password('wrong', 'new'), false);
});

test('adds payment method', () => {
  const acc = new Account('a@b.com', 'secret');
  const count = acc.add_payment_method('Visa-1234');
  assert.strictEqual(count, 1);
});

test('removes payment method', () => {
  const acc = new Account('a@b.com', 'secret');
  acc.add_payment_method('Visa-1234');
  const result = acc.remove_payment_method(0);
  assert.deepStrictEqual(result, []);
});

test('sets notification preference', () => {
  const acc = new Account('a@b.com', 'secret');
  const result = acc.set_notification_preference('sms', true);
  assert.strictEqual(result.sms, true);
});

test('exports all data', () => {
  const acc = new Account('a@b.com', 'secret');
  const result = acc.export_data();
  assert.strictEqual(result.email, 'a@b.com');
  assert.ok(result.auditLog);
});

test('logs access', () => {
  const acc = new Account('a@b.com', 'secret');
  const count = acc.log_access('view');
  assert.strictEqual(count, 1);
});

test('checks subscription', () => {
  const acc = new Account('a@b.com', 'secret');
  assert.strictEqual(acc.check_subscription(), 'basic');
});

test('upgrades subscription', () => {
  const acc = new Account('a@b.com', 'secret');
  const result = acc.upgrade_subscription('pro');
  assert.strictEqual(result, 'pro');
});

File: python/12-god-object/test/account_test.py
from src.account import Account

def test_login_correct_password():
    acc = Account("a@b.com", "secret")
    assert acc.login("secret") is True

def test_reject_wrong_password():
    acc = Account("a@b.com", "secret")
    assert acc.login("wrong") is False

def test_logout():
    acc = Account("a@b.com", "secret")
    assert acc.logout() is True

def test_update_profile():
    acc = Account("a@b.com", "secret")
    result = acc.update_profile("Alice", "Developer")
    assert result["name"] == "Alice"
    assert result["bio"] == "Developer"

def test_change_password():
    acc = Account("a@b.com", "secret")
    assert acc.change_password("secret", "new") is True
    assert acc.login("new") is True

def test_reject_bad_old_password():
    acc = Account("a@b.com", "secret")
    assert acc.change_password("wrong", "new") is False

def test_add_payment_method():
    acc = Account("a@b.com", "secret")
    count = acc.add_payment_method("Visa-1234")
    assert count == 1

def test_remove_payment_method():
    acc = Account("a@b.com", "secret")
    acc.add_payment_method("Visa-1234")
    result = acc.remove_payment_method(0)
    assert result == []

def test_set_notification_preference():
    acc = Account("a@b.com", "secret")
    result = acc.set_notification_preference("sms", True)
    assert result["sms"] is True

def test_export_all_data():
    acc = Account("a@b.com", "secret")
    result = acc.export_data()
    assert result["email"] == "a@b.com"
    assert "auditLog" in result

def test_log_access():
    acc = Account("a@b.com", "secret")
    count = acc.log_access("view")
    assert count == 1

def test_check_subscription():
    acc = Account("a@b.com", "secret")
    assert acc.check_subscription() == "basic"

def test_upgrade_subscription():
    acc = Account("a@b.com", "secret")
    result = acc.upgrade_subscription("pro")
    assert result == "pro"
